File: go/pkg/genx/transformers/doubao_tts_icl_v2.go
package transformers

import (
	"context"
	"io"
	"strings"

	"github.com/haivivi/giztoy/go/pkg/doubaospeech"
	"github.com/haivivi/giztoy/go/pkg/genx"
)

// DoubaoTTSICLV2 is a TTS transformer using Doubao seed-icl-2.0 (声音复刻 2.0).
//
// Resource ID: seed-icl-2.0
//
// Speaker: Custom cloned voice ID starting with "S_" (e.g., "S_xxxxxx")
//
// Input type: text/plain
// Output type: audio/* (audio/ogg by default)
//
// EoS Handling:
//   - When receiving a text/plain EoS marker, finish synthesis, emit audio chunks, then emit audio/* EoS
//   - Non-text chunks are passed through unchanged
type DoubaoTTSICLV2 struct {
	client      *doubaospeech.Client
	speaker     string
	format      string
	sampleRate  int
	bitRate     int
	speedRatio  float64
	volumeRatio float64
	pitchRatio  float64
	emotion     string
	language    string
}

var _ genx.Transformer = (*DoubaoTTSICLV2)(nil)

// DoubaoTTSICLV2Option is a functional option for DoubaoTTSICLV2.
type DoubaoTTSICLV2Option func(*DoubaoTTSICLV2)

// WithDoubaoTTSICLV2Format sets the audio format (pcm, mp3, ogg_opus).
func WithDoubaoTTSICLV2Format(format string) DoubaoTTSICLV2Option {
	return func(t *DoubaoTTSICLV2) {
		t.format = format
	}
}

// WithDoubaoTTSICLV2SampleRate sets the sample rate (8000, 16000, 24000, 32000).
func WithDoubaoTTSICLV2SampleRate(sampleRate int) DoubaoTTSICLV2Option {
	return func(t *DoubaoTTSICLV2) {
		t.sampleRate = sampleRate
	}
}

// WithDoubaoTTSICLV2BitRate sets the bit rate for mp3 (32000, 64000, 128000).
func WithDoubaoTTSICLV2BitRate(bitRate int) DoubaoTTSICLV2Option {
	return func(t *DoubaoTTSICLV2) {
		t.bitRate = bitRate
	}
}

// WithDoubaoTTSICLV2Speed sets the speech speed ratio (0.2-3.0).
func WithDoubaoTTSICLV2Speed(speedRatio float64) DoubaoTTSICLV2Option {
	return func(t *DoubaoTTSICLV2) {
		t.speedRatio = speedRatio
	}
}

// WithDoubaoTTSICLV2Volume sets the volume ratio (0.1-3.0).
func WithDoubaoTTSICLV2Volume(volumeRatio float64) DoubaoTTSICLV2Option {
	return func(t *DoubaoTTSICLV2) {
		t.volumeRatio = volumeRatio
	}
}

// WithDoubaoTTSICLV2Pitch sets the pitch ratio (0.1-3.0).
func WithDoubaoTTSICLV2Pitch(pitchRatio float64) DoubaoTTSICLV2Option {
	return func(t *DoubaoTTSICLV2) {
		t.pitchRatio = pitchRatio
	}
}

// WithDoubaoTTSICLV2Emotion sets the emotion (happy, sad, angry, fear, hate, surprise).
func WithDoubaoTTSICLV2Emotion(emotion string) DoubaoTTSICLV2Option {
	return func(t *DoubaoTTSICLV2) {
		t.emotion = emotion
	}
}

// WithDoubaoTTSICLV2Language sets the language (zh, en, ja, etc.).
func WithDoubaoTTSICLV2Language(language string) DoubaoTTSICLV2Option {
	return func(t *DoubaoTTSICLV2) {
		t.language = language
	}
}

// NewDoubaoTTSICLV2 creates a new DoubaoTTSICLV2 transformer.
//
// Parameters:
//   - client: Doubao speech client
//   - speaker: Custom voice ID (must start with "S_", e.g., "S_xxxxxx")
//   - opts: Optional configuration
func NewDoubaoTTSICLV2(client *doubaospeech.Client, speaker string, opts ...DoubaoTTSICLV2Option) *DoubaoTTSICLV2 {
	t := &DoubaoTTSICLV2{
		client:      client,
		speaker:     speaker,
		format:      "ogg_opus",
		sampleRate:  24000,
		speedRatio:  1.0,
		volumeRatio: 1.0,
		pitchRatio:  1.0,
	}
	for _, opt := range opts {
		opt(t)
	}
	return t
}

// DoubaoTTSICLV2CtxKey is the context key for runtime options.
type doubaoTTSICLV2CtxKey struct{}

// DoubaoTTSICLV2CtxOptions are runtime options passed via context.
// TODO: Add fields as needed for runtime configuration.
type DoubaoTTSICLV2CtxOptions struct{}

// WithDoubaoTTSICLV2CtxOptions attaches runtime options to context.
func WithDoubaoTTSICLV2CtxOptions(ctx context.Context, opts DoubaoTTSICLV2CtxOptions) context.Context {
	return context.WithValue(ctx, doubaoTTSICLV2CtxKey{}, opts)
}

// Transform converts Text chunks to audio Blob chunks.
// DoubaoTTSICLV2 does not require connection setup, so it returns immediately.
// The ctx is unused (no initialization needed); the goroutine lifetime
// is governed by the input Stream.
func (t *DoubaoTTSICLV2) Transform(_ context.Context, _ string, input genx.Stream) (genx.Stream, error) {
	output := newBufferStream(100)

	go t.transformLoop(input, output)

	return output, nil
}

func (t *DoubaoTTSICLV2) transformLoop(input genx.Stream, output *bufferStream) {
	defer output.Close()

	// Local cancel context tied to the loop lifecycle.
	// When the loop exits, defer cancel() cancels any in-flight HTTP request.
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	mimeType := t.mimeType()
	var textBuilder strings.Builder
	var lastChunk *genx.MessageChunk

	for {
		chunk, err := input.Next()
		if err != nil {
			if err != io.EOF {
				output.CloseWithError(err)
				return
			}
			// EOF: synthesize any remaining text
			if textBuilder.Len() > 0 {
				if err := t.synthesize(ctx, textBuilder.String(), lastChunk, mimeType, output); err != nil {
					output.CloseWithError(err)
					return
				}
			}
			return
		}

		if chunk == nil {
			continue
		}

		lastChunk = chunk

		// Check for text EoS marker
		if chunk.IsEndOfStream() {
			if _, ok := chunk.Part.(genx.Text); ok {
				// Text EoS: synthesize accumulated text, emit audio, then emit audio EoS
			if textBuilder.Len() > 0 {
				if err := t.synthesize(ctx, textBuilder.String(), lastChunk, mimeType, output); err != nil {
					output.CloseWithError(err)
					return
				}
				textBuilder.Reset()
			}
				// Emit audio EoS
				eosChunk := genx.NewEndOfStream(mimeType)
				if lastChunk != nil {
					eosChunk.Role = lastChunk.Role
					eosChunk.Name = lastChunk.Name
				}
				if err := output.Push(eosChunk); err != nil {
					return
				}
				continue
			}
			// Non-text EoS: pass through
			if err := output.Push(chunk); err != nil {
				return
			}
			continue
		}

		// Collect text
		if text, ok := chunk.Part.(genx.Text); ok {
			textBuilder.WriteString(string(text))
		} else {
			// Non-text chunk: pass through
			if err := output.Push(chunk); err != nil {
				return
			}
		}
	}
}

func (t *DoubaoTTSICLV2) synthesize(ctx context.Context, text string, lastChunk *genx.MessageChunk, mimeType string, output *bufferStream) error {
	req := &doubaospeech.TTSV2Request{
		Text:        text,
		Speaker:     t.speaker,
		ResourceID:  doubaospeech.ResourceVoiceCloneV2,
		Format:      t.format,
		SampleRate:  t.sampleRate,
		BitRate:     t.bitRate,
		SpeedRatio:  t.speedRatio,
		VolumeRatio: t.volumeRatio,
		PitchRatio:  t.pitchRatio,
		Emotion:     t.emotion,
		Language:    t.language,
	}

	for chunk, err := range t.client.TTSV2.Stream(ctx, req) {
		if err != nil {
			return err
		}

		if chunk.Audio != nil && len(chunk.Audio) > 0 {
			outChunk := &genx.MessageChunk{
				Part: &genx.Blob{
					MIMEType: mimeType,
					Data:     chunk.Audio,
				},
			}
			if lastChunk != nil {
				outChunk.Role = lastChunk.Role
				outChunk.Name = lastChunk.Name
			}

			if err := output.Push(outChunk); err != nil {
				return err
			}
		}
	}
	return nil
}

func (t *DoubaoTTSICLV2) mimeType() string {
	switch t.format {
	case "mp3":
		return "audio/mpeg"
	case "ogg_opus":
		return "audio/ogg"
	case "pcm":
		return "audio/pcm"
	default:
		return "audio/ogg"
	}
}

File: rust/genx/src/transformers/doubao_tts_icl_v2.rs
use std::sync::Arc;

use async_trait::async_trait;
use giztoy_doubaospeech as doubaospeech;

use crate::error::GenxError;
use crate::stream::Stream;
use crate::transformer::Transformer;

use super::doubao_tts_seed_v2::DoubaoTtsV2Provider;
use super::tts_core::{spawn_tts_transform_loop, TtsProvider};

pub struct DoubaoTtsIclV2Transformer {
    provider: Arc<dyn TtsProvider>,
}

impl DoubaoTtsIclV2Transformer {
    pub fn new(client: Arc<doubaospeech::Client>, speaker: impl Into<String>) -> Self {
        Self {
            provider: Arc::new(DoubaoTtsV2Provider {
                client,
                speaker: speaker.into(),
                resource_id: doubaospeech::RESOURCE_VOICE_CLONE_V2.to_string(),
                format: "ogg_opus".to_string(),
                sample_rate: Some(24000),
                bit_rate: None,
                speed_ratio: Some(1.0),
                volume_ratio: Some(1.0),
                pitch_ratio: Some(1.0),
                emotion: None,
                language: None,
            }),
        }
    }

    pub(crate) fn new_with_provider(provider: Arc<dyn TtsProvider>) -> Self {
        Self { provider }
    }
}

#[async_trait]
impl Transformer for DoubaoTtsIclV2Transformer {
    async fn transform(
        &self,
        _pattern: &str,
        input: Box<dyn Stream>,
    ) -> Result<Box<dyn Stream>, GenxError> {
        Ok(spawn_tts_transform_loop(Arc::clone(&self.provider), input))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::transformers::tts_core::AudioEmitter;
    use crate::types::{MessageChunk, Role};
    use tokio::sync::Mutex;

    struct MockProvider {
        responses: Mutex<Vec<Result<Vec<Vec<u8>>, GenxError>>>,
    }

    #[async_trait]
    impl TtsProvider for MockProvider {
        fn mime_type(&self) -> &str {
            "audio/ogg"
        }

        async fn synthesize_stream(
            &self,
            _text: &str,
            emitter: &mut dyn AudioEmitter,
        ) -> Result<(), GenxError> {
            match self.responses.lock().await.remove(0) {
                Ok(chunks) => {
                    for audio in chunks {
                        emitter.emit(audio).await?;
                    }
                    Ok(())
                }
                Err(e) => Err(e),
            }
        }
    }

    fn input_stream(chunks: Vec<MessageChunk>) -> Box<dyn Stream> {
        let builder = crate::stream::StreamBuilder::with_tools(16, vec![]);
        builder.add(&chunks).expect("add chunks");
        builder.done(crate::error::Usage::default()).expect("done");
        Box::new(builder.stream())
    }

    #[tokio::test]
    async fn tts_error_chain_is_propagated() {
        let provider: Arc<dyn TtsProvider> = Arc::new(MockProvider {
            responses: Mutex::new(vec![Err(GenxError::Other(anyhow::anyhow!(
                "icl provider error"
            )))]),
        });
        let t = DoubaoTtsIclV2Transformer::new_with_provider(provider);

        let mut out = t
            .transform(
                "doubao/icl-v2",
                input_stream(vec![
                    MessageChunk::text(Role::Model, "test"),
                    MessageChunk::new_text_end_of_stream(),
                ]),
            )
            .await
            .expect("transform");

        let _ = out.next().await.expect("bos").expect("bos chunk");
        let err = out.next().await.expect_err("provider error");
        assert!(err.to_string().contains("icl provider error"));
    }
}
